lib/trino-parquet/src/main/java/io/trino/parquet/reader/FilteredOffsetIndex.java

```diff
@@ -0,0 +1,163 @@
+
+    @Override
+    public long getOffset(int pageIndex)
+    {
+        return offsetIndex.getOffset(indexMap[pageIndex]);
+    }
+
+    @Override
+    public int getCompressedPageSize(int pageIndex)
+    {
+        return offsetIndex.getCompressedPageSize(indexMap[pageIndex]);
+    }
+
+    @Override
+    public long getFirstRowIndex(int pageIndex)
+    {
+        return offsetIndex.getFirstRowIndex(indexMap[pageIndex]);
+    }
+
+    @Override
+    public long getLastRowIndex(int pageIndex, long totalRowCount)
+    {
+        int nextIndex = indexMap[pageIndex] + 1;
+        return (nextIndex >= offsetIndex.getPageCount() ? totalRowCount : offsetIndex.getFirstRowIndex(nextIndex)) - 1;
+    }
+
+    @Override
+    public String toString()
+    {
+        try (Formatter formatter = new Formatter()) {
+            formatter.format("%-12s  %20s  %16s  %20s\n", "", "offset", "compressed size", "first row index");
+            for (int i = 0, n = offsetIndex.getPageCount(); i < n; ++i) {
+                int index = Arrays.binarySearch(indexMap, i);
+                boolean isHidden = index < 0;
+                formatter.format("%spage-%-5d  %20d  %16d  %20d\n",
+                        isHidden ? "- " : "  ",
+                        isHidden ? i : index,
+                        offsetIndex.getOffset(i),
+                        offsetIndex.getCompressedPageSize(i),
+                        offsetIndex.getFirstRowIndex(i));
+            }
+            return formatter.toString();
+        }
+    }
+
+    List<OffsetRange> calculateOffsetRanges(long rowGroupOffset)
+    {
+        List<OffsetRange> ranges = new ArrayList<>();
+        int n = this.getPageCount();
+        if (n > 0) {
+            OffsetRange currentRange = null;
+
+            long firstPageOffset = this.offsetIndex.getOffset(0);
+            // Add a range for the dictionary page if required
+            if (rowGroupOffset < firstPageOffset) {
+                // We need to adjust the offset by startingPosition for presto because dataSource.readFully() started at startingPosition
+                currentRange = new OffsetRange(rowGroupOffset /*- startingPosition*/, (int) (firstPageOffset - rowGroupOffset));
+                ranges.add(currentRange);
+            }
+
+            for (int i = 0; i < n; ++i) {
+                long offset = this.getOffset(i);
+                int length = this.getCompressedPageSize(i);
+                // We need to adjust the offset by startingPosition for presto because dataSource.readFully() started at startingPosition
+                if (currentRange == null || !currentRange.extend(offset /*- startingPosition*/, length)) {
+                    currentRange = new OffsetRange(offset /*- startingPosition*/, length);
+                    ranges.add(currentRange);
+                }
+            }
+        }
+        return ranges;
+    }
+
+    static class OffsetRange
+    {
+        private final long offset;
+        private long length;
```

Comment:
It'd be safer and easier to read the code if this class were immutable, especially since values of OffsetRange are returned from the `calculateOffsetRanges` method.

This will require re-working the method above. Something like:

```
            ...
            long currentOffset = -1;
            long currentLength = 0;

            long firstPageOffset = offsetIndex.getOffset(0);
            // Add a range for the dictionary page if required
            if (rowGroupOffset < firstPageOffset) {
                // We need to adjust the offset by startingPosition for trino because dataSource.readFully() started at startingPosition
                currentOffset = rowGroupOffset;
                currentLength = firstPageOffset - rowGroupOffset;
            }

            for (int i = 0; i < pageCount; ++i) {
                long offset = getOffset(i);
                int length = getCompressedPageSize(i);

                if (currentOffset == -1) {
                    currentOffset = offset;
                    currentLength = length;
                }
                else if (currentOffset + currentLength == offset) {
                    currentLength += length;
                }
                else {
                    ranges.add(new OffsetRange(currentOffset, currentLength));
                    currentOffset = -1;
                    currentLength = 0;
                }
            }

            ranges.add(new OffsetRange(currentOffset, currentLength));
            
            ...
```